pkg/referrerstore/oras/cosign.go

```diff
@@ -16,71 +16,63 @@ limitations under the License.
 package oras
 
 import (
-	"crypto/tls"
+	"context"
 	"errors"
+	"fmt"
 	"net/http"
 	"strings"
 
 	"github.com/deislabs/ratify/pkg/common"
 	"github.com/deislabs/ratify/pkg/ocispecs"
-	"github.com/google/go-containerregistry/pkg/name"
-	v1 "github.com/google/go-containerregistry/pkg/v1"
-	"github.com/google/go-containerregistry/pkg/v1/remote"
-	"github.com/google/go-containerregistry/pkg/v1/remote/transport"
-	"github.com/opencontainers/go-digest"
 	oci "github.com/opencontainers/image-spec/specs-go/v1"
+
+	"oras.land/oras-go/v2/errdef"
+	"oras.land/oras-go/v2/registry"
+	"oras.land/oras-go/v2/registry/remote/errcode"
 )
 
-const CosignArtifactType = "org.sigstore.cosign.v1"
+const CosignArtifactType = "application/vnd.dev.cosign.artifact.sig.v1+json"
 const CosignSignatureTagSuffix = ".sig"
 
-func getCosignReferences(subjectReference common.Reference, config *OrasStoreConf) (*[]ocispecs.ReferenceDescriptor, error) {
+var ErrNoCosignSubjectDigest = errors.New("failed to mutate cosign image tag: no digest specified for subject")
+
+func getCosignReferences(ctx context.Context, subjectReference common.Reference, store *orasStore, repository registry.Repository) (*[]ocispecs.ReferenceDescriptor, error) {
 	var references []ocispecs.ReferenceDescriptor
-	var opts []name.Option
-	var remoteOptions []remote.Option
-	if isInsecureRegistry(subjectReference.Original, config) {
-		opts = append(opts, name.Insecure)
-		// #nosec G402
-		remoteOptions = append(remoteOptions, remote.WithTransport(&http.Transport{TLSClientConfig: &tls.Config{InsecureSkipVerify: true}}))
-	}
-	ref, err := name.ParseReference(subjectReference.Original, opts...)
+	signatureTag, err := attachedImageTag(subjectReference, CosignSignatureTagSuffix)
 	if err != nil {
-		return &references, err
+		return nil, err
 	}
-	hash := v1.Hash{
-		Algorithm: subjectReference.Digest.Algorithm().String(),
-		Hex:       subjectReference.Digest.Hex(),
-	}
-
-	signatureTag := attachedImageTag(ref.Context(), hash, CosignSignatureTagSuffix)
 
-	desc, err := remote.Get(signatureTag, remoteOptions...)
-	var terr *transport.Error
+	desc, err := repository.Resolve(ctx, signatureTag)
 	if err != nil {
-		if errors.As(err, &terr) && terr.StatusCode == http.StatusNotFound {
-			return &references, nil
+		if errors.Is(err, errdef.ErrNotFound) {
+			return nil, nil
 		}
-		return &references, err
-	}
-	descDig, err := digest.Parse(desc.Digest.String())
-	if err != nil {
-		return &references, err
+		var ec errcode.Error
+		if errors.As(err, &ec) && (ec.Code == fmt.Sprint(http.StatusForbidden) || ec.Code == fmt.Sprint(http.StatusUnauthorized)) {
+			store.evictAuthCache(subjectReference.Original, err)
```

Comment:
wonder if we should also evict cache on `ErrNotFound`?

Reply:
The `ErrNotFound` will be returned by ORAS if the manifest doesn't exist. We're using that currently to do the existence check to see if there are any cosign sigs. My understanding is that any auth related issues would return a 401 or 403 only.

Reply:
Just realized it's auth cache, ignore my comment.